File: Tests/Ray_vs_straight/shootrays.m
function [rx,rz,Dr] = shootrays(p, v_profile, zmax, dr)
% [rx,rz,Dr] = ray_trace_1D(p, v_profile, zmax, dr)
% 
% 1D ray tracing function

vdz = 5;

zz1 = [0:vdz:(zmax-vdz)]';
zz2 = [vdz:vdz:zmax]';

v1  = linterp(v_profile.z,v_profile.vwater,zz1);
v2  = linterp(v_profile.z,v_profile.vwater,zz2);
% assume flat Earth (no radius terms)
u1 = 1./v1;
u2 = 1./v2; 

dv = v2-v1;

dz = zz2-zz1;
b = dv ./ dz;
const_indx = (b == 0);

X = zeros( length(v1), 1 );
X(const_indx) = constv_dist( u1(const_indx), dz(const_indx), p );
X(~const_indx) = gradv_dist(b(~const_indx),u1(~const_indx),u2(~const_indx),p);
X = [0; X];
X(imag(X)>0) = NaN;
X = X(~isnan(X));
Xd = cumsum(X);
Xd = r2d(Xd/6371); % convert km to degrees
[rlat,rlon] = reckon( slat, slon, Xd, baz );
[rayx,rayy] = project_xy( par, rlat, rlon );
rayz = [zz1(1); zz2]; 
rayz = rayz(1:length(X));
dz = dz(1:length(X)-1);

% calc Dr
r_scale = (1/6371).*(6371-rayz);
X2 = r_scale.*X;
Dr1 = (X2(2:end).^2 + dz.^2).^(1/2);
Dr1 = [0; Dr1];
Dr2 = cumsum(Dr1);
Dr = [0:dr:max(Dr2)]';
Dr = [Dr; max(Dr2)];
% cDr = Dr;
rx = interp1(Dr2,rayx,Dr);
ry = interp1(Dr2,rayy,Dr);
rz = interp1(Dr2,rayz,Dr);
% nn = 1:round(length(Dr)/length(rz)):length(Dr)+10;
% nn = nn(1:length(rz));
% Dr = interp1(1:length(Dr),Dr,nn)'; % this will be exactly the same Dr as above
% cDr = interp1(1:length(cDr),cDr,nn)'; % this is the same as Dr


%-------------------- sub-functions --------------------------------------
%
function x = constv_dist( u, dz, p )

x = (p * dz) ./ eta(u,p);


function x = gradv_dist( b, u1, u2, p )

x1 = eta(u1,p) ./ (b.*u1*p);
x2 = eta(u2,p) ./ (b.*u2*p);

x = x1 - x2;

%
function val = eta( u, p )

val = sqrt( u.^2 - p.^2 );

File: OBSrange_v1_PYTHON/funcs/shootrays.py
'''
FUNCTION shootrays.py 

1D ray tracing function all units in km or km/s

Z. Eilon 01/2019 (modified from code by Brandon Schmandt)
Translated to Python and further modified by S. Mosher 01/2019
'''
# Import modules and functions
import numpy as np

from IPython.core.debugger import Tracer


# Given two arrays, this function finds the closest value in the smaller array
# to each value in the larger array.
def find_ilay(r, R):
  
  N = len(r)
  ilay = np.zeros(N, dtype=int)
  
  for i, depth in enumerate(r):
    idx = np.abs(R - depth).argmin()
    ilay[i] = idx

  return ilay

# Linear interpolation function.
def linterp(X, Y, XI):
  
  YI = np.zeros(len(XI))

  ilay = find_ilay(XI, X)
  
  # Linearly interpolate.
  YI = (XI - X[ilay]) * (Y[ilay + 1] - Y[ilay]) / (X[ilay + 1] - X[ilay]) + Y[ilay]

  # Sort out coincident elements.
  olap = np.intersect1d(X, XI)
  for i, val in enumerate(olap):
    YI[XI == olap[i]] = np.mean(Y[X == olap[i]])

  return YI



def shootrays(p, v_profile, zmax, dr=0.001, vdz=0.001):

  # Anonymous functions
  eta = lambda u, p: np.sqrt(u**2 + p**2)
  gradv_dist = lambda b, u1, u2, p: (eta(u1,p)/(b*u1*p)) - (eta(u2,p)/(b*u2*p))
  constv_dist = lambda u, dz, p: (p * dz) / eta(u, p)

  zz1 = np.arange(0, zmax, vdz)
  zz2 = np.arange(vdz, zmax+vdz, vdz)
  

  v1  = linterp(v_profile[0], v_profile[1], zz1)
  v2  = linterp(v_profile[0], v_profile[1], zz2)
  
  Tracer()()

  # assume flat Earth (no radius terms)
  u1 = 1 / v1
  u2 = 1 / v2 
  
  dv = v2 - v1
  
  dz = zz2 - zz1
  b = dv / dz
  const_indx = (b == 0)
  
  X = np.zeros(len(v1))
  X[const_indx == True] = constv_dist( u1[const_indx == True], dz[const_indx == True], p )
  X[const_indx != True] = gradv_dist(b[const_indx != True], u1[const_indx != True],u2[const_indx != True],p)
  X = np.hstack([0, X])
  X[np.imag(X) > 0] = np.nan
  X = X[X != np.isnan(X)]
  Xd = np.cumsum(X)
  rayz = np.hstack([zz1[1], zz2]) 
  #rayz = rayz(1:length(X))
  #dz = dz(1:length(X)-1)
  
  # calc Dr
  Dr1 = np.sqrt(X**2 + dz**2)
  assert(np.imag(Dr1).any() == False)
  Dr1 = np.hstack([0, Dr1])
  Dr2 = np.cumsum(Dr1)
  #Dr = np.unique( np.hstack( [np.linspace(0, max(Dr2) + dr, dr), max(Dr2)])).T
  #rx = interp1(Dr2,Xd,Dr)
  #rz = interp1(Dr2,rayz,Dr)
  #
  #% calculate travel time (JBR)
  #dDr = Dr(2:length(Dr))-Dr(1:length(Dr)-1)
  #dDr = [0; dDr]
  #v = [v1(1; v2)]
  #rv = interp1(Dr2,v,Dr)
  #dt = dDr./rv
  #rt = cumsum(dt)

  #return 